js/main.js

```diff
@@ -29,58 +29,145 @@ Site = {
       maxZoom: 20,
       id: 'mapbox.streets'
     }).addTo(_this.map);
+  },
 
-    var coordinateArray = [];
+  getDevicePos: function(callback) {
+    var _this = this;
 
-    _this.map.on('mouseup', function() {
+    navigator.geolocation.getCurrentPosition(function(position) {
+      _this.pos.lat = position.coords.latitude;
+      _this.pos.long = position.coords.longitude;
 
-      _this.map.off("mousemove");
+      callback();
+    });
+  },
 
-      L.polyline(coordinateArray, {color: '#f94000'}).addTo(_this.map);
+  flyToDevicePos: function() {
+    var _this = this;
 
-      console.log(coordinateArray);
+    _this.getDevicePos(function() {
+      _this.map.flyTo([_this.pos.lat, _this.pos.long], 20, {
+        animate: true
+      });
 
-      coordinateArray = [];
+      _this.stateDraw();
+      _this.bindStates();
+    });
+  },
 
-      navigator.geolocation.getCurrentPosition(function(position) {
-        _this.map.panTo(new L.LatLng(position.coords.latitude, position.coords.longitude));
-      });
-     
+  draw: function() {
+    var _this = this;
 
+    _this.map.on('mouseup', function() {
+      _this.map.off("mousemove");
+      L.polyline(_this.coordinateArray, {color: 'black'}).addTo(_this.map);
+      _this.coordinateArray = [];
     }).on('mousedown', function() {
-
       _this.map.on("mousemove", function(e) {
-        trackPoints(e)
+        _this.trackMouse(e);
       });
+    });
+  },
 
+  trackMouse: function(e) {
+    var _this = this;
+
+    _this.coordinateArray.push([e.latlng.lat, e.latlng.lng]);
+  },
+
+  stateDraw: function() {
+    var _this = this;
+
+    $('#draw').prop('disabled', true);
+    $('#move').prop('disabled', false);
+    _this.map.locate({
+      watch: true,
+      setView: true
     });
+    _this.moveDisable();
+    _this.draw();
+  },
 
-    function trackPoints(e) {
-      coordinateArray.push([e.latlng.lat, e.latlng.lng]);
-    }
+  stateMove: function() {
+    var _this = this;
+    
+    $('#draw').prop('disabled', false);
+    $('#move').prop('disabled', true);
+    _this.map.stopLocate();
+    _this.moveEnable();
+    _this.map.off('mouseup mousedown');
+  },
+
+  moveDisable: function() {
+    var _this = this;
+
+    _this.map.dragging.disable();
+    _this.map.touchZoom.disable();
+    _this.map.doubleClickZoom.disable();
+    _this.map.scrollWheelZoom.disable();
+    _this.map.boxZoom.disable();
+    _this.map.keyboard.disable();
+    _this.map.zoomControl.disable();
   },
-}
+
+  moveEnable: function() {
+    var _this = this;
+
+    _this.map.dragging.enable();
+    _this.map.touchZoom.enable();
+    _this.map.doubleClickZoom.enable();
+    _this.map.scrollWheelZoom.enable();
+    _this.map.boxZoom.enable();
+    _this.map.keyboard.enable();
+    _this.map.zoomControl.enable();
+  },
+
+  bindStates: function() {
+    var _this = this;
+
+    $('#move').on('click', function() {
+      _this.stateMove();
```

Comment:
I think the concept of state doesn't work as well unless there is something reactive to that state. As in the state setting and changing can trigger other events.